monitoring/perf_context.cc

```diff
@@ -38,177 +38,185 @@ PerfContext::~PerfContext() {
-    env_reuse_writable_file_nanos = other.env_reuse_writable_file_nanos;
-    env_new_random_rw_file_nanos = other.env_new_random_rw_file_nanos;
-    env_new_directory_nanos = other.env_new_directory_nanos;
-    env_file_exists_nanos = other.env_file_exists_nanos;
-    env_get_children_nanos = other.env_get_children_nanos;
-    env_get_children_file_attributes_nanos = other.env_get_children_file_attributes_nanos;
-    env_delete_file_nanos = other.env_delete_file_nanos;
-    env_create_dir_nanos = other.env_create_dir_nanos;
-    env_create_dir_if_missing_nanos = other.env_create_dir_if_missing_nanos;
-    env_delete_dir_nanos = other.env_delete_dir_nanos;
-    env_get_file_size_nanos = other.env_get_file_size_nanos;
-    env_get_file_modification_time_nanos = other.env_get_file_modification_time_nanos;
-    env_rename_file_nanos = other.env_rename_file_nanos;
-    env_link_file_nanos = other.env_link_file_nanos;
-    env_lock_file_nanos = other.env_lock_file_nanos;
-    env_unlock_file_nanos = other.env_unlock_file_nanos;
-    env_new_logger_nanos = other.env_new_logger_nanos;
-    get_cpu_nanos = other.get_cpu_nanos;
-    if (other.per_level_perf_context_enabled &&
-        other.level_to_perf_context != nullptr) {
-      ClearPerLevelPerfContext();
+  env_new_sequential_file_nanos = other.env_new_sequential_file_nanos;
+  env_new_random_access_file_nanos = other.env_new_random_access_file_nanos;
+  env_new_writable_file_nanos = other.env_new_writable_file_nanos;
+  env_reuse_writable_file_nanos = other.env_reuse_writable_file_nanos;
+  env_new_random_rw_file_nanos = other.env_new_random_rw_file_nanos;
+  env_new_directory_nanos = other.env_new_directory_nanos;
+  env_file_exists_nanos = other.env_file_exists_nanos;
+  env_get_children_nanos = other.env_get_children_nanos;
+  env_get_children_file_attributes_nanos = other.env_get_children_file_attributes_nanos;
+  env_delete_file_nanos = other.env_delete_file_nanos;
+  env_create_dir_nanos = other.env_create_dir_nanos;
+  env_create_dir_if_missing_nanos = other.env_create_dir_if_missing_nanos;
+  env_delete_dir_nanos = other.env_delete_dir_nanos;
+  env_get_file_size_nanos = other.env_get_file_size_nanos;
+  env_get_file_modification_time_nanos = other.env_get_file_modification_time_nanos;
+  env_rename_file_nanos = other.env_rename_file_nanos;
+  env_link_file_nanos = other.env_link_file_nanos;
+  env_lock_file_nanos = other.env_lock_file_nanos;
+  env_unlock_file_nanos = other.env_unlock_file_nanos;
+  env_new_logger_nanos = other.env_new_logger_nanos;
+  get_cpu_nanos = other.get_cpu_nanos;
+  if (per_level_perf_context_enabled && level_to_perf_context != nullptr) {
+    ClearPerLevelPerfContext();
+  }
+  if (other.level_to_perf_context != nullptr) {
+    if (other.per_level_perf_context_enabled) {
       EnablePerLevelPerfContext();
-      *level_to_perf_context = *other.level_to_perf_context;
     }
-  #endif
+    *level_to_perf_context = *other.level_to_perf_context;
```

Comment:
How do we make sure `level_to_perf_context` is not null?

Reply:
the assumption is that when other.level_to_perf_context is not null, other.per_level_perf_context_enabled will always be true. I guess I can drop the `if (other.per_level_perf_context_enabled) {` check in line 122 and always call `EnablePerLevelPerfContext();` before doing the assignment for `*level_to_perf_context`?

Reply:
This way, the sequence I mentioned above won't work, which is not good.

Reply:
added test coverage for the sequence above